### ISSUE TITLE
Softmax and big data

### ISSUE THREAD

**Issue by rolfans** [Microsoft/CNTK#2728]
Hi, I am testing some semantic segmentation in CNTK, and when I add softmax for the loss function the net gets really slow, it is not able to utilize the GPU any more, without the softmax it uses 90-100% of the GPU, with it only 15-20. What is going on? My images are 512x512.

**Comment by KeDengMS** [comment]
Can you please share a repro?

**Comment by rolfans** [comment]
No, but the problem is like this:
z=makeNetwork(512,512)
z1=C.softmax(z) # without this softmax it uses about 2 hours pr iteration and about 100% of my GPU, with it is uses 12 hours and about 15% of the GPU.

**Comment by KeDengMS** [comment]
We need more details to investigate this issue, or you may build CNTK from source and debug into SoftmaxNode.

**Comment by rolfans** [comment]
The base is the unet segmetnation: https://github.com/usuyama/cntk_unet (which example does not work if you use a bigger image), the resnet fcn https://github.com/usuyama/cntk-fully-convolutional-networks , my deeplab v2 converted from Caffe, and some other not completed conversion test. My own image reader functions to get my images in, and a lot of tests on different loss and error functions.  And all the nets slow down when I add the softmax to then.

**Comment by KeDengMS** [comment]
Thanks for the info. Does the slow down happen in inference, or training? What's the image size, label dimension, and batch size?

**Comment by rolfans** [comment]
In training, the image size is 512x512. The batch size is dependent on the network, how many images it can take into memory of the Titan XP card, so from 4 and up. The label images have the same dimensions

**Comment by KeDengMS** [comment]
So are you running Softmax on a 512x512 dimension? That might be slow in the current implementation of SoftmaxNode. Can you try this instead?

    i = C.input_variable((512,512,))
    m = C.reduce_max(i)
    sm = C.exp(i - m - C.reduce_log_sum_exp(i - m))

**Comment by rolfans** [comment]
At first I though it would make it slower, since the GPU load was 0-2% for a long time before a small 99% peak, long 0-2% small peak 99% and so on. But It was actually much faster. 
So thank you.
